Refactor code to simplify manifest items and their creation

gsnedders opened the pull request.

gsnedders commented on tools/manifest/item.py:
```diff
@@ -108,6 +88,10 @@ def id(self):
 
+    @property
+    def url_base(self):
+        return "/"
```
This is blatantly wrong. But also currently doesn't break anything (uh, we should have tests!).

gsnedders replied:
Hah, seemingly this is actually a constant! (We always set it to `/` and `SourceFile` expands the full URL.) And only used by `.url`, so we can just drop this.

jgraham replied:
Note that it's set by the manifest and doenstream projects (notably gecko and servo) have manifests which set values other than `/` e.g. https://searchfox.org/mozilla-central/source/testing/web-platform/mozilla/meta/MANIFEST.json#1768

jgraham replied:
(That might not mean this code wasn't useless, just checking you're aware of where the url base is used).

jgraham commented on tools/manifest/item.py:
```diff
@@ -52,12 +37,7 @@ def id(self):
-    def path(self):
-        """The test path relative to the test_root"""
-        return self.source_file.rel_path
+        return set()
```
Is the belief that this isn't used anywhere? If so can we add a warning or something if we don't want to remove it?

gsnedders replied:
Yes. AFAICT, it is literally only used by `ManifestItem.https` and nowhere else, and we don't need to deal with that for cases where we don't have a URL. I guess `https` should really move to `URLManifestItem`, and `meta_flags` made private?

gsnedders committed:
Make manifest items leaf nodes with no dependencies

jgraham commented on tools/wptrunner/wptrunner/wpttest.py:
```diff
@@ -149,12 +149,12 @@ def update_metadata(self, metadata=None):
                    test_metadata,
                    timeout=timeout,
-                   path=manifest_item.source_file.path,
+                   path=os.path.join(manifest_file.tests_root, manifest_item.path),
```
You could have a method on the manifest to return this

gsnedders replied:
That seems a bit weird to me, at least.

gsnedders commented:
So, despite this mostly being done as refactoring to do other work for #15137, this has made the single-test case (`time ./wpt run --binary=`which chromium` --headless --no-pause chrome infrastructure/assumptions/html-elements.html`) 20% faster on my desktop computer.

gsnedders commented on tools/wptrunner/wptrunner/testloader.py:
```diff
@@ -504,7 +504,7 @@ def iter_tests(self):
 
         for test_type, test_path, tests in manifest_items:
-            manifest_file = manifests_by_url_base[iter(tests).next().url_base]
+            manifest_file = manifests_by_url_base["/"]
```
This is obviously wrong. However, given `item.url_base` was previously just the constant `"/"`, this isn't a regression. I presume @jgraham will say I need to fix this code though. )

jgraham replied:
You need to fix this code. Or change the iterator to produce something different like `(manifest, source_file, tests)`.

gsnedders replied:
So, we only get this right when loading the JSON, not when actually creating the manifest.

Ms2ger commented on tools/wptrunner/wptrunner/wpttest.py:
```diff
@@ -313,32 +313,28 @@ class TestharnessTest(Test):
 
         self.testdriver = testdriver
         self.jsshell = jsshell
-        self.scripts = scripts or []
```
I do need *some* way to get these; they're used in https://searchfox.org/mozilla-central/rev/cb7faaf6b4ad2528390186f1ce64618dea71031e/js/src/tests/jstests.py#404

gsnedders replied:
@jgraham told me it was unused! :)

Guess we need to add it to the JSON, then.

jgraham replied:
Well it's not that expensive to read from the file. But if the goal is that we don't read any test file data except metadata when the test file hasn't been updated, then yes.

gsnedders replied:
> But if the goal is that we don't read any test file data except metadata when the test file hasn't been updated, then yes.

We shouldn't need to read any test file data to just run the test.

gsnedders replied:
So I've fixed this by putting it in the manifest… but I fear this means we're now gonna be broken until the MANIFEST is regenerated from scratch given it won't contain this until it's regenerated…

jgraham replied:
We can bump the manifest version number to force it to be regenerated. We should probably also make sure that all the manifest upload things regenerate from scratch every time.

gsnedders replied:
Following on from discussion on IRC, I've re-added `tests_root` to items (yay memory consumption!) and made it possible for them to create `SourceFile` objects (albeit behind a private property) to fetch the `scripts` from there if the `_extras` doesn't already contain it.

gsnedders replied:
And, to be clear, when we next bump the MANIFEST version we should eliminate all of that.

gsnedders committed:
Create items from SourceFile with correct url_base

gsnedders committed:
Remove the EqualTimeChunker

This hasn't been used by anyone in several years, and trying to fix
the tests as part of the work for #13507 tends to suggest that the
chunker is itself broken. It is, in many ways, the least useful
chunker as it results in test execution order changing every time the
test execution time metadata is updated, and this leads to further
flakiness.

gsnedders commented:
Can't figure out why the `test_wpttest.py` tests are failing; will need to dig in deeper next week.

gsnedders commented:
The one remaining failure is in `test_list_tests_missing_manifest`. This appears to be caused by us passing `--tests tools/wpt/tests` into the `wpt run` call in that test, and this hella confuses things. We end up generating a manifest for the git repo root but then plenty of other places aren't aware we're doing this.

gsnedders commented:
Given @Hexcles mentioned memory consumption of manifest generation in https://github.com/web-platform-tests/wpt/issues/12874#issuecomment-465794369:

On Linux/x86_64, on this branch, manifest generation from scratch consumes max 487MB, clean regeneration 202MB.

`master`, by way of comparison requires max 1408MB and 290MB respectively.

Hexcles commented:
That is definitely promising! I also tested the memory footprint of manifest before (~1.5G from scratch) and it is indeed a problem for folks running tests on VMs (or Taskcluster workers, from forks of this repo) with <=4G memory.

We should also try to free the memory ASAP. How much memory does the manifest take when we are running the tests (without manifest generation)? I assume it'd take take less than generating the manifest, and if so, it'd be great to make sure the delta is freed when the generation is completed.

gsnedders commented:
And the commit I've just pushed here makes it go down to max 165MB and 187MB respectively (the delta I presume is caused by it loading the old manifest and therefore having that in memory too).

But this does now mean we aren't holding on to any file beyond the point at which we're directly processing that file.

jgraham reviewed (commented):
THis looks great. Let me double check it doesn't break gecko before approving.

jgraham commented:
It seems like this broke gecko by calling git unconditionally https://treeherder.mozilla.org/#/jobs?repo=try&author=james%40hoppipolla.co.uk&collapsedPushes=420652&fromchange=551ebec95f9b6ad9b5ac3d8812ee04b61d9024c3&selectedJob=230362281

jgraham reviewed (changes requested):
Need to ensure that `git` isn't required when updating with `--work`.

gsnedders commented:
> Need to ensure that `git` isn't required when updating with `--work`.

That's fixed now.

gsnedders commented on tools/manifest/item.py:
```diff
@@ -38,39 +22,26 @@ def __new__(cls, name, bases, attrs, **kwargs):
         """A unique identifier for the test"""
         return (self.item_type, self.id)
 
     def meta_key(self):
```
I… don't think so? But I don't really have the confidence here. We represent items as sets in various places which means we rely on the hashes to some extent, at least.

jgraham replied:
Sure, but unless we are (still) using the hash to determine whether a test updated or not then we can probably remove this method. Not for this PR anyway.

gsnedders committed:
Get rid of tests for behaviour that's impossible

gsnedders committed:
Fix TestharnessTest to have a correct meta_key

gsnedders committed:
SourceFile always generated RefTestNode, never RefTest

In various places we write tests under the assumption that SourceFile
returns the applicable type, which it doesn't. Fix our tests to always
return RefTestNode. This allows us to get rid of a try/except block
that exists purely to handle this case which is unreachable without
mocks.

gsnedders committed:
new_type can never be falsy, just use an else block

foolip commented:
@gsnedders I've been running a script to regenerate manifests for all tagged commits locally and compare with what can be downloaded from GitHub releases. Starting with merge_pr_15308 there are a lot that have an added slash at the beginning of paths. I'm not sure why that happened, was that ever the intended behavior before, during or after this change?

foolip commented:
@gsnedders using "dom/nodes/Document-doctype.html" as a test path that's been around since 2014, it looks like the leading slash was originally not there in the manifest and at some point was added. From which commit should it be there, when doing a full rebuild of the manifest?

gsnedders commented:
@foolip not sure; it shouldn't be there (and isn't today).

foolip commented:
@gsnedders alright, I'll work on the assumption that it should never have been there. So far it looks like it's there for some manifests created using an incremental update, but not if you recreate the manifest from scratch on the same commit.
